api/server/router/system/system_routes.go

```diff
@@ -55,10 +57,24 @@ func (s *systemRouter) getEvents(ctx context.Context, w http.ResponseWriter, r *
 		return err
 	}
 
-	var timeout <-chan time.Time
+	var (
+		timeout        <-chan time.Time
+		onlyPastEvents bool
+	)
 	if until > 0 || untilNano > 0 {
-		dur := time.Unix(until, untilNano).Sub(time.Now())
-		timeout = time.NewTimer(dur).C
+		if since > until {
+			return errors.NewBadRequestError(fmt.Errorf("`since` time (%s) cannot be after `until` time (%s)", r.Form.Get("since"), r.Form.Get("until")))
+		}
```

Comment:
I think this comparison should include the fractional parts as well. Probably the cleanest way to do it is to convert the values to `time.Time` and then use the time package's comparison functions. Then if we go with https://github.com/docker/docker/pull/21939/files#r59301234, we could pass these `time.Time` values to `SubscribeToEvents`.